File: go/PopulateNextRightPointer/PopulateNextRightPointer.go
package main

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

// Time -> 0(N) Space -> 0(N)
func populate(tree *Node) *Node {
	queue := make([][]*Node, 0)
	queue = append(queue, []*Node{tree})
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		next := make([]*Node, 0)
		for len(node) > 0 {
			curr := node[0]
			node = node[1:]

			if curr == nil {
				continue
			}

			if len(node) > 0 {
				curr.Next = node[0]
			}

			next = append(next, curr.Left)
			next = append(next, curr.Right)
		}

		if len(next) > 0 {
			queue = append(queue, next)
		}
	}
	return tree
}

File: python/populateNextRightPointer.py
class Node:
    def __init__(self, val=0, left=None, right=None, next=None) -> None:
        self.val = val
        self.left = left
        self.right = right
        self.next = next


def populate(root: Node | None) -> Node | None:
    queue = []
    queue.append([root])
    while len(queue) > 0:
        node = queue.pop(0)
        next = []
        while len(node) > 0:
            curr = node.pop(0)

            if curr is None:
                continue

            if len(node) > 0:
                curr.next = node[0]

            next.append(curr.left)
            next.append(curr.right)

        if len(next) > 0:
            queue.append(next)

    return root
